### PR TITLE
[Enhancement] Add Dev tools to boost develop

### PR DESCRIPTION
**Pull request by sennnnn**
### Add more dev tools to boost develop progress (Mainly for **regression benchmark**)

Goals:
1. url check to find out invalid checkpoint url (log json url is contained but it's hacky.).
2. more convenient model gather tool.
2. test model inference of important model list.
3. test model evaluation of important model list.
4. test model training of important model list.

TODO:
- [x] url check tool: `.dev/check_urls.py`;
- [x] more convenient model gather tool: `.dev/gather_models.py`;
- [x] model-ckpt list and model_config list: `.dev/batch_test_list.py` & `batch_train_list.txt`;
- [x] model inference test tool: `.dev/benchmark_inference.py`;
- [x] model evaluation test tool script and script generator: `.dev/benchmark_evaluation.sh` & `generate_benchmark_evaluation_script.py`;
- [x] gather evaluation results and compare evaluation results tool: `.dev/gather_benchmark_evaluation_results.py`;
- [x] model training test tool scipt and script generator: `.dev/benchmark_train.sh` & `generate_benchmark_train_script.py`;
- [x] gather training results and compare training results tool: `.dev/gather_benchmark_train_results.py`;

### CODE REVIEW

**Comment by sennnnn** [comment]
# A glance of benchmark model lists

```python
hrnet = [
    dict(
        config='configs/hrnet/fcn_hr18s_512x512_160k_ade20k.py',
        checkpoint='fcn_hr18s_512x512_160k_ade20k_20200614_214413-870f65ac.pth', # noqa
        eval='mIoU',
        metric=dict(mIoU=33.0),
    ),
    dict(
        config='configs/hrnet/fcn_hr18s_512x1024_160k_cityscapes.py',
        checkpoint='fcn_hr18s_512x1024_160k_cityscapes_20200602_190901-4a0797ea.pth', # noqa
        eval='mIoU',
        metric=dict(mIoU=76.31),
    ),
    dict(
        config='configs/hrnet/fcn_hr48_512x512_160k_ade20k.py',
        checkpoint='fcn_hr48_512x512_160k_ade20k_20200614_214407-a52fc02c.pth',
        eval='mIoU',
        metric=dict(mIoU=42.02),
    ),
    dict(
        config='configs/hrnet/fcn_hr48_512x1024_160k_cityscapes.py',
        checkpoint='fcn_hr48_512x1024_160k_cityscapes_20200602_190946-59b7973e.pth', # noqa
        eval='mIoU',
        metric=dict(mIoU=80.65),
    ),
]
pspnet = [
    dict(
        config='configs/pspnet/pspnet_r50-d8_512x1024_80k_cityscapes.py',
        checkpoint='pspnet_r50-d8_512x1024_80k_cityscapes_20200606_112131-2376f12b.pth', # noqa
        eval='mIoU',
        metric=dict(mIoU=78.55),
    ),
    dict(
        config='configs/pspnet/pspnet_r101-d8_512x1024_80k_cityscapes.py',
        checkpoint='pspnet_r101-d8_512x1024_80k_cityscapes_20200606_112211-e1e1100f.pth', # noqa
        eval='mIoU',
        metric=dict(mIoU=79.76),
    ),
    dict(
        config='configs/pspnet/pspnet_r101-d8_512x512_160k_ade20k.py',
        checkpoint='pspnet_r101-d8_512x512_160k_ade20k_20200615_100650-967c316f.pth', # noqa
        eval='mIoU',
        metric=dict(mIoU=44.39),
    ),
    dict(
        config='configs/pspnet/pspnet_r50-d8_512x512_160k_ade20k.py',
        checkpoint='pspnet_r50-d8_512x512_160k_ade20k_20200615_184358-1890b0bd.pth', # noqa
        eval='mIoU',
        metric=dict(mIoU=42.48),
    ),
]
resnest = [
    dict(
        config='configs/resnest/pspnet_s101-d8_512x512_160k_ade20k.py',
        checkpoint='pspnet_s101-d8_512x512_160k_ade20k_20200807_145416-a6daa92a.pth', # noqa
        eval='mIoU',
        metric=dict(mIoU=45.44),
    ),
    dict(
        config='configs/resnest/pspnet_s101-d8_512x1024_80k_cityscapes.py',
        checkpoint='pspnet_s101-d8_512x1024_80k_cityscapes_20200807_140631-c75f3b99.pth', # noqa
        eval='mIoU',
        metric=dict(mIoU=78.57),
    ),
]
fastscnn = [
    dict(
        config='configs/fastscnn/fast_scnn_lr0.12_8x4_160k_cityscapes.py',
        checkpoint='fast_scnn_8x4_160k_lr0.12_cityscapes-0cec9937.pth',
        eval='mIoU',
        metric=dict(mIoU=70.96),
    )
]
deeplabv3plus = [
    dict(
        config='configs/deeplabv3plus/deeplabv3plus_r101-d8_769x769_80k_cityscapes.py', # noqa
        checkpoint='deeplabv3plus_r101-d8_769x769_80k_cityscapes_20200607_000405-a7573d20.pth', # noqa
        eval='mIoU',
        metric=dict(mIoU=80.98),
    ),
    dict(
        config='configs/deeplabv3plus/deeplabv3plus_r101-d8_512x1024_80k_cityscapes.py', # noqa
        checkpoint='deeplabv3plus_r101-d8_512x1024_80k_cityscapes_20200606_114143-068fcfe9.pth', # noqa
        eval='mIoU',
        metric=dict(mIoU=80.97),
    ),
    dict(
        config='configs/deeplabv3plus/deeplabv3plus_r50-d8_512x1024_80k_cityscapes.py', # noqa
        checkpoint='deeplabv3plus_r50-d8_512x1024_80k_cityscapes_20200606_114049-f9fb496d.pth', # noqa
        eval='mIoU',
        metric=dict(mIoU=80.09),
    ),
    dict(
        config='configs/deeplabv3plus/deeplabv3plus_r50-d8_769x769_80k_cityscapes.py', # noqa
        checkpoint='deeplabv3plus_r50-d8_769x769_80k_cityscapes_20200606_210233-0e9dfdc4.pth', # noqa
        eval='mIoU',
        metric=dict(mIoU=79.83),
    ),
]
vit = [
    dict(
        config='configs/vit/upernet_vit-b16_ln_mln_512x512_160k_ade20k.py',
        checkpoint='upernet_vit-b16_ln_mln_512x512_160k_ade20k-f444c077.pth',
        eval='mIoU',
        metric=dict(mIoU=47.73),
    ),
    dict(
        config='configs/vit/upernet_deit-s16_ln_mln_512x512_160k_ade20k.py',
        checkpoint='upernet_deit-s16_ln_mln_512x512_160k_ade20k-c0cd652f.pth',
        eval='mIoU',
        metric=dict(mIoU=43.52),
    ),
]
fp16 = [
    dict(
        config='configs/fp16/deeplabv3plus_r101-d8_512x1024_80k_fp16_cityscapes.py', # noqa
        checkpoint='deeplabv3plus_r101-d8_512x1024_80k_fp16_cityscapes-cc58bc8d.pth', # noqa
        eval='mIoU',
        metric=dict(mIoU=80.46),
    )
]
swin = [
    dict(
        config='configs/swin/upernet_swin_tiny_patch4_window7_512x512_160k_ade20k_pretrain_224x224_1K.py', # noqa
        checkpoint='upernet_swin_tiny_patch4_window7_512x512_160k_ade20k_pretrain_224x224_1K_20210531_112542-e380ad3e.pth', # noqa
        eval='mIoU',
        metric=dict(mIoU=44.41),
    )
]
```

**Comment by sennnnn** [comment]
Regression Benchmark Evaluation Results:
fcn_hr18s_512x512_160k_ade20k & fast_scnn_lr0.12_8x4_160k_cityscapes missing
```json
{
    "configs/hrnet/fcn_hr18s_512x1024_160k_cityscapes.py": {
        "differential": {
            "mIoU": "+0.02"
        },
        "previous": {
            "mIoU": 76.31
        },
        "new": {
            "mIoU": 76.33
        }
    },
    "configs/hrnet/fcn_hr48_512x512_160k_ade20k.py": {
        "differential": {
            "mIoU": "-0.00"
        },
        "previous": {
            "mIoU": 42.02
        },
        "new": {
            "mIoU": 42.02
        }
    },
    "configs/hrnet/fcn_hr48_512x1024_160k_cityscapes.py": {
        "differential": {
            "mIoU": "+0.02"
        },
        "previous": {
            "mIoU": 80.65
        },
        "new": {
            "mIoU": 80.67
        }
    },
    "configs/pspnet/pspnet_r50-d8_512x1024_80k_cityscapes.py": {
        "differential": {
            "mIoU": "+0.11"
        },
        "previous": {
            "mIoU": 78.55
        },
        "new": {
            "mIoU": 78.66
        }
    },
    "configs/pspnet/pspnet_r101-d8_512x1024_80k_cityscapes.py": {
        "differential": {
            "mIoU": "-0.02"
        },
        "previous": {
            "mIoU": 79.76
        },
        "new": {
            "mIoU": 79.74
        }
    },
    "configs/pspnet/pspnet_r101-d8_512x512_160k_ade20k.py": {
        "differential": {
            "mIoU": "-0.00"
        },
        "previous": {
            "mIoU": 44.39
        },
        "new": {
            "mIoU": 44.39
        }
    },
    "configs/pspnet/pspnet_r50-d8_512x512_160k_ade20k.py": {
        "differential": {
            "mIoU": "-0.00"
        },
        "previous": {
            "mIoU": 42.48
        },
        "new": {
            "mIoU": 42.48
        }
    },
    "configs/resnest/pspnet_s101-d8_512x512_160k_ade20k.py": {
        "differential": {
            "mIoU": "-0.00"
        },
        "previous": {
            "mIoU": 45.44
        },
        "new": {
            "mIoU": 45.44
        }
    },
    "configs/resnest/pspnet_s101-d8_512x1024_80k_cityscapes.py": {
        "differential": {
            "mIoU": "-0.00"
        },
        "previous": {
            "mIoU": 78.57
        },
        "new": {
            "mIoU": 78.57
        }
    },
    "configs/deeplabv3plus/deeplabv3plus_r101-d8_769x769_80k_cityscapes.py": {
        "differential": {
            "mIoU": "-0.06"
        },
        "previous": {
            "mIoU": 80.98
        },
        "new": {
            "mIoU": 80.92
        }
    },
    "configs/deeplabv3plus/deeplabv3plus_r101-d8_512x1024_80k_cityscapes.py": {
        "differential": {
            "mIoU": "+0.07"
        },
        "previous": {
            "mIoU": 80.97
        },
        "new": {
            "mIoU": 81.04
        }
    },
    "configs/deeplabv3plus/deeplabv3plus_r50-d8_512x1024_80k_cityscapes.py": {
        "differential": {
            "mIoU": "-0.08"
        },
        "previous": {
            "mIoU": 80.09
        },
        "new": {
            "mIoU": 80.01
        }
    },
    "configs/deeplabv3plus/deeplabv3plus_r50-d8_769x769_80k_cityscapes.py": {
        "differential": {
            "mIoU": "-0.00"
        },
        "previous": {
            "mIoU": 79.83
        },
        "new": {
            "mIoU": 79.83
        }
    },
    "configs/vit/upernet_vit-b16_ln_mln_512x512_160k_ade20k.py": {
        "differential": {
            "mIoU": "+1.04"
        },
        "previous": {
            "mIoU": 47.73
        },
        "new": {
            "mIoU": 48.77
        }
    },
    "configs/vit/upernet_deit-s16_ln_mln_512x512_160k_ade20k.py": {
        "differential": {
            "mIoU": "+0.67"
        },
        "previous": {
            "mIoU": 43.52
        },
        "new": {
            "mIoU": 44.19
        }
    },
    "configs/fp16/deeplabv3plus_r101-d8_512x1024_80k_fp16_cityscapes.py": {
        "differential": {
            "mIoU": "-0.00"
        },
        "previous": {
            "mIoU": 80.46
        },
        "new": {
            "mIoU": 80.46
        }
    },
    "configs/swin/upernet_swin_tiny_patch4_window7_512x512_160k_ade20k_pretrain_224x224_1K.py": {
        "differential": {
            "mIoU": "-0.00"
        },
        "previous": {
            "mIoU": 44.41
        },
        "new": {
            "mIoU": 44.41
        }
    }
}
```

**Comment by sennnnn** [comment]
Regression Benchmark Training Results:
```json
{
    "configs/hrnet/fcn_hr18s_512x512_160k_ade20k.py": {
        "mIoU": 32.9,
        "mAcc": 43.1,
        "aAcc": 76.9
    },
    "configs/hrnet/fcn_hr18s_512x1024_160k_cityscapes.py": {
        "mIoU": 76.6,
        "mAcc": 84.5,
        "aAcc": 96.0
    },
    "configs/hrnet/fcn_hr48_512x512_160k_ade20k.py": {
        "mIoU": 42.0,
        "mAcc": 53.6,
        "aAcc": 80.2
    },
    "configs/hrnet/fcn_hr48_512x1024_160k_cityscapes.py": {
        "mIoU": 81.0,
        "mAcc": 87.9,
        "aAcc": 96.7
    },
    "configs/pspnet/pspnet_r50-d8_512x1024_80k_cityscapes.py": {
        "mIoU": 78.4,
        "mAcc": 85.2,
        "aAcc": 96.2
    },
    "configs/pspnet/pspnet_r101-d8_512x1024_80k_cityscapes.py": {
        "mIoU": 80.1,
        "mAcc": 87.1,
        "aAcc": 96.4
    },
    "configs/pspnet/pspnet_r101-d8_512x512_160k_ade20k.py": {
        "mIoU": 43.8,
        "mAcc": 53.8,
        "aAcc": 81.2
    },
    "configs/pspnet/pspnet_r50-d8_512x512_160k_ade20k.py": {
        "mIoU": 42.4,
        "mAcc": 52.2,
        "aAcc": 80.6
    },
    "configs/resnest/pspnet_s101-d8_512x512_160k_ade20k.py": {
        "mIoU": 44.6,
        "mAcc": 55.8,
        "aAcc": 81.2
    },
    "configs/fastscnn/fast_scnn_lr0.12_8x4_160k_cityscapes.py": {
        "mIoU": 71.0,
        "mAcc": 78.8,
        "aAcc": 94.9
    },
    "configs/deeplabv3plus/deeplabv3plus_r101-d8_769x769_80k_cityscapes.py": {
        "mIoU": 81.0,
        "mAcc": 88.8,
        "aAcc": 96.5
    },
    "configs/deeplabv3plus/deeplabv3plus_r101-d8_512x1024_80k_cityscapes.py": {
        "mIoU": 81.1,
        "mAcc": 87.8,
        "aAcc": 96.6
    },
    "configs/deeplabv3plus/deeplabv3plus_r50-d8_512x1024_80k_cityscapes.py": {
        "mIoU": 80.6,
        "mAcc": 87.9,
        "aAcc": 96.5
    },
    "configs/deeplabv3plus/deeplabv3plus_r50-d8_769x769_80k_cityscapes.py": {
        "mIoU": 80.1,
        "mAcc": 88.4,
        "aAcc": 96.5
    },
    "configs/vit/upernet_vit-b16_ln_mln_512x512_160k_ade20k.py": {
        "mIoU": 48.1,
        "mAcc": 58.9,
        "aAcc": 82.6
    },
    "configs/vit/upernet_deit-s16_ln_mln_512x512_160k_ade20k.py": {
        "mIoU": 44.2,
        "mAcc": 55.2,
        "aAcc": 81.1
    },
    "configs/fp16/deeplabv3plus_r101-d8_512x1024_80k_fp16_cityscapes.py": {
        "mIoU": 80.7,
        "mAcc": 87.4,
        "aAcc": 96.5
    },
    "configs/swin/upernet_swin_tiny_patch4_window7_512x512_160k_ade20k_pretrain_224x224_1K.py": {
        "mIoU": 44.0,
        "mAcc": 55.0,
        "aAcc": 81.0
    }
}
```

**Comment by Junjun2016** [comment]
> Regression Benchmark Training Results:
> 
> ```json
> {
>     "configs/hrnet/fcn_hr18s_512x512_160k_ade20k.py": {
>         "mIoU": 32.9,
>         "mAcc": 43.1,
>         "aAcc": 76.9
>     },
>     "configs/hrnet/fcn_hr18s_512x1024_160k_cityscapes.py": {
>         "mIoU": 76.6,
>         "mAcc": 84.5,
>         "aAcc": 96.0
>     },
>     "configs/hrnet/fcn_hr48_512x512_160k_ade20k.py": {
>         "mIoU": 42.0,
>         "mAcc": 53.6,
>         "aAcc": 80.2
>     },
>     "configs/hrnet/fcn_hr48_512x1024_160k_cityscapes.py": {
>         "mIoU": 81.0,
>         "mAcc": 87.9,
>         "aAcc": 96.7
>     },
>     "configs/pspnet/pspnet_r50-d8_512x1024_80k_cityscapes.py": {
>         "mIoU": 78.4,
>         "mAcc": 85.2,
>         "aAcc": 96.2
>     },
>     "configs/pspnet/pspnet_r101-d8_512x1024_80k_cityscapes.py": {
>         "mIoU": 80.1,
>         "mAcc": 87.1,
>         "aAcc": 96.4
>     },
>     "configs/pspnet/pspnet_r101-d8_512x512_160k_ade20k.py": {
>         "mIoU": 43.8,
>         "mAcc": 53.8,
>         "aAcc": 81.2
>     },
>     "configs/pspnet/pspnet_r50-d8_512x512_160k_ade20k.py": {
>         "mIoU": 42.4,
>         "mAcc": 52.2,
>         "aAcc": 80.6
>     },
>     "configs/resnest/pspnet_s101-d8_512x512_160k_ade20k.py": {
>         "mIoU": 44.6,
>         "mAcc": 55.8,
>         "aAcc": 81.2
>     },
>     "configs/fastscnn/fast_scnn_lr0.12_8x4_160k_cityscapes.py": {
>         "mIoU": 71.0,
>         "mAcc": 78.8,
>         "aAcc": 94.9
>     },
>     "configs/deeplabv3plus/deeplabv3plus_r101-d8_769x769_80k_cityscapes.py": {
>         "mIoU": 81.0,
>         "mAcc": 88.8,
>         "aAcc": 96.5
>     },
>     "configs/deeplabv3plus/deeplabv3plus_r101-d8_512x1024_80k_cityscapes.py": {
>         "mIoU": 81.1,
>         "mAcc": 87.8,
>         "aAcc": 96.6
>     },
>     "configs/deeplabv3plus/deeplabv3plus_r50-d8_512x1024_80k_cityscapes.py": {
>         "mIoU": 80.6,
>         "mAcc": 87.9,
>         "aAcc": 96.5
>     },
>     "configs/deeplabv3plus/deeplabv3plus_r50-d8_769x769_80k_cityscapes.py": {
>         "mIoU": 80.1,
>         "mAcc": 88.4,
>         "aAcc": 96.5
>     },
>     "configs/vit/upernet_vit-b16_ln_mln_512x512_160k_ade20k.py": {
>         "mIoU": 48.1,
>         "mAcc": 58.9,
>         "aAcc": 82.6
>     },
>     "configs/vit/upernet_deit-s16_ln_mln_512x512_160k_ade20k.py": {
>         "mIoU": 44.2,
>         "mAcc": 55.2,
>         "aAcc": 81.1
>     },
>     "configs/fp16/deeplabv3plus_r101-d8_512x1024_80k_fp16_cityscapes.py": {
>         "mIoU": 80.7,
>         "mAcc": 87.4,
>         "aAcc": 96.5
>     },
>     "configs/swin/upernet_swin_tiny_patch4_window7_512x512_160k_ade20k_pretrain_224x224_1K.py": {
>         "mIoU": 44.0,
>         "mAcc": 55.0,
>         "aAcc": 81.0
>     }
> }
> ```

Are the results in this file equal to the results of `new` in the above file?

**Comment by Junjun2016** [comment]
It seems not for some items.

**Comment by Junjun2016** [comment]
Hi @sennnnn 
Please fix the lint error.
`update-model-index` hook has been fixed.

File: .dev/gather_benchmark_train_results.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
+import argparse
+import glob
+import os.path as osp
+
+import mmcv
+from gather_models import get_final_results
+from mmcv import Config
```
Comment (Junjun2016) [inline .dev/gather_benchmark_train_results.py:7]: Does `get_final_results` work for `xxx.log.json` of the latest version?
```
{"mode": "train", "epoch": 430, "iter": 159800, "lr": 0.00012, "memory": 18550, "data_time": 0.03098, "decode.final.loss_seg": 0.1267, "decode.final.acc_seg": 89.37481, "decode.4x.loss_seg": 0.05068, "decode.4x.acc_seg": 89.37481, "decode.cascade_1.loss_seg": 0.05473, "decode.cascade_1.acc_seg": 89.02916, "decode.cascade_2.loss_seg": 0.06245, "decode.cascade_2.acc_seg": 88.45142, "loss": 0.29455, "time": 0.40958}
{"mode": "train", "epoch": 430, "iter": 159850, "lr": 0.00012, "memory": 18550, "data_time": 0.01244, "decode.final.loss_seg": 0.12426, "decode.final.acc_seg": 90.30909, "decode.4x.loss_seg": 0.0497, "decode.4x.acc_seg": 90.30909, "decode.cascade_1.loss_seg": 0.05396, "decode.cascade_1.acc_seg": 89.95941, "decode.cascade_2.loss_seg": 0.06163, "decode.cascade_2.acc_seg": 89.38377, "loss": 0.28955, "time": 0.45076}
{"mode": "train", "epoch": 430, "iter": 159900, "lr": 0.00011, "memory": 18550, "data_time": 0.01909, "decode.final.loss_seg": 0.12876, "decode.final.acc_seg": 89.90795, "decode.4x.loss_seg": 0.0515, "decode.4x.acc_seg": 89.90795, "decode.cascade_1.loss_seg": 0.05563, "decode.cascade_1.acc_seg": 89.55699, "decode.cascade_2.loss_seg": 0.06348, "decode.cascade_2.acc_seg": 88.96176, "loss": 0.29938, "time": 0.20294}
{"mode": "train", "epoch": 430, "iter": 159950, "lr": 0.00011, "memory": 18550, "data_time": 0.04161, "decode.final.loss_seg": 0.12413, "decode.final.acc_seg": 90.28908, "decode.4x.loss_seg": 0.04965, "decode.4x.acc_seg": 90.28908, "decode.cascade_1.loss_seg": 0.0539, "decode.cascade_1.acc_seg": 89.94204, "decode.cascade_2.loss_seg": 0.06162, "decode.cascade_2.acc_seg": 89.37222, "loss": 0.2893, "time": 0.48495}
{"mode": "train", "epoch": 431, "iter": 160000, "lr": 0.0001, "memory": 18550, "data_time": 0.02879, "decode.final.loss_seg": 0.12311, "decode.final.acc_seg": 90.08211, "decode.4x.loss_seg": 0.04924, "decode.4x.acc_seg": 90.08211, "decode.cascade_1.loss_seg": 0.05334, "decode.cascade_1.acc_seg": 89.74749, "decode.cascade_2.loss_seg": 0.06098, "decode.cascade_2.acc_seg": 89.16612, "loss": 0.28667, "time": 0.34189}
{"mode": "val", "epoch": 431, "iter": 125, "lr": 0.0001, "aAcc": 0.9527, "mIoU": 0.727, "mAcc": 0.8069, "IoU.road": 0.9799, "IoU.sidewalk": 0.8352, "IoU.building": 0.9073, "IoU.wall": 0.4933, "IoU.fence": 0.5505, "IoU.pole": 0.5568, "IoU.traffic light": 0.6097, "IoU.traffic sign": 0.7214, "IoU.vegetation": 0.9115, "IoU.terrain": 0.5975, "IoU.sky": 0.9395, "IoU.person": 0.7605, "IoU.rider": 0.5405, "IoU.car": 0.9367, "IoU.truck": 0.759, "IoU.bus": 0.816, "IoU.train": 0.6353, "IoU.motorcycle": 0.5443, "IoU.bicycle": 0.7187, "Acc.road": 0.9881, "Acc.sidewalk": 0.9236, "Acc.building": 0.9546, "Acc.wall": 0.5448, "Acc.fence": 0.6566, "Acc.pole": 0.6921, "Acc.traffic light": 0.7166, "Acc.traffic sign": 0.802, "Acc.vegetation": 0.9651, "Acc.terrain": 0.6985, "Acc.sky": 0.9758, "Acc.person": 0.8721, "Acc.rider": 0.7069, "Acc.car": 0.9719, "Acc.truck": 0.8161, "Acc.bus": 0.8869, "Acc.train": 0.6681, "Acc.motorcycle": 0.6356, "Acc.bicycle": 0.8555}
```

**Review by Junjun2016 — COMMENTED** [review]
Please resolve comments.

File: .dev/gather_benchmark_train_results.py (hunk 1 above)
Comment (sennnnn) [reply .dev/gather_benchmark_train_results.py:7]: Fixed.
You can refer to this [line](https://github.com/sennnnn/mmsegmentation/blob/27866dcbc0df938cec35a8aa311e1440ad34b498/.dev/gather_models.py#L63)

File: .dev/check_urls.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                logger.error(
+                    f'checkpoint | {config_name} | {checkpoint_url} | '
+                    f'{status_code} | error')
+            # log_json check
+            checkpoint_name = checkpoint_url.split('/')[-1]
+            model_time = '-'.join(checkpoint_name.split('-')[:-1]).replace(
+                f'{config_name}_', '')
+            # two style of log_json name
+            # use '_' to link model_time (will be deprecated)
+            log_json_url_1 = f'https://download.openmmlab.com/mmsegmentation/v0.5/{model_name}/{config_name}/{config_name}_{model_time}.log.json'  # noqa
```
[the remark this answers is not in the document]
Comment (sennnnn) [reply .dev/check_urls.py:81]: Yes, log url should be recorded in metafiles.

File: .dev/check_urls.py [hunk 3]
```diff
@@ -21,9 +21,7 @@ def check_url(url):
     flag = True
     r = requests.head(url)
     status_code = r.status_code
-    if status_code == 403:
-        flag = False
-    elif status_code == 404:
+    if status_code == 403 or 404:
```
Comment (Junjun2016) [inline .dev/check_urls.py:24]:
```suggestion
    if status_code == 403 or status_code == 404:
```

**Review by Junjun2016 — APPROVED** [review]
LGTM

File: tools/train.py [hunk 4]
```diff
@@ -77,8 +77,10 @@ def main():
         cfg.work_dir = args.work_dir
     elif cfg.get('work_dir', None) is None:
         # use config filename as default work_dir if cfg.work_dir is None
-        cfg.work_dir = osp.join('./work_dirs',
-                                osp.splitext(osp.basename(args.config))[0])
+        config_dir = osp.split(osp.dirname(args.config))[1]
+        config_name = osp.splitext(osp.basename(args.config))[0]
+        cfg.work_dir = osp.join('./work_dirs', config_dir, config_name)
```
Comment (xvjiarui) [inline tools/train.py:82]: Please take the following case into consideration:

The user added 1 config file under the folder `configs/`, or directory under the root of repo. `config_dir` will become `configs` or `MMSegmentation`, which is not idea.

Comment (xvjiarui) [reply tools/train.py:82]: I suggest keep this part unchanged.

File: tools/test.py [hunk 5]
```diff
@@ -21,6 +23,9 @@ def parse_args():
         description='mmseg test (and eval) a model')
     parser.add_argument('config', help='test config file path')
     parser.add_argument('checkpoint', help='checkpoint file')
+    parser.add_argument(
+        '--work-dir',
+        help='the directory to save the file containing evaluation metrics')
```
Comment (xvjiarui) [inline tools/test.py:28]:
```suggestion
        help='if specified, the evaluation metric results will be dumped into the directory as json')
```

File: .dev/benchmark_evaluation.sh [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+echo 'configs/fp16/deeplabv3plus_r101-d8_512x1024_80k_fp16_cityscapes.py' &
+GPUS=4  GPUS_PER_NODE=4  CPUS_PER_TASK=2 tools/slurm_test.sh $PARTITION deeplabv3plus_r101-d8_512x1024_80k_fp16_cityscapes configs/fp16/deeplabv3plus_r101-d8_512x1024_80k_fp16_cityscapes.py $CHECKPOINT_DIR/deeplabv3plus_r101-d8_512x1024_80k_fp16_cityscapes-cc58bc8d.pth --eval mIoU --work-dir .dev/benchmark_test/deeplabv3plus_r101-d8_512x1024_80k_fp16_cityscapes --options dist_params.port=28188 &
+echo 'configs/swin/upernet_swin_tiny_patch4_window7_512x512_160k_ade20k_pretrain_224x224_1K.py' &
+GPUS=4  GPUS_PER_NODE=4  CPUS_PER_TASK=2 tools/slurm_test.sh $PARTITION upernet_swin_tiny_patch4_window7_512x512_160k_ade20k_pretrain_224x224_1K configs/swin/upernet_swin_tiny_patch4_window7_512x512_160k_ade20k_pretrain_224x224_1K.py $CHECKPOINT_DIR/upernet_swin_tiny_patch4_window7_512x512_160k_ade20k_pretrain_224x224_1K_20210531_112542-e380ad3e.pth --eval mIoU --work-dir .dev/benchmark_test/upernet_swin_tiny_patch4_window7_512x512_160k_ade20k_pretrain_224x224_1K --options dist_params.port=28189 &
```
Comment (xvjiarui) [inline .dev/benchmark_evaluation.sh:41]: We may specify `work_dir` for these scripts.

File: .dev/benchmark_train.sh [hunk 7]
```diff
@@ -1,40 +1,40 @@
 PARTITION=$1
 
 echo 'configs/hrnet/fcn_hr18s_512x512_160k_ade20k.py' &
-GPUS=4  GPUS_PER_NODE=4  CPUS_PER_TASK=2 ./tools/slurm_train.sh $PARTITION fcn_hr18s_512x512_160k_ade20k configs/hrnet/fcn_hr18s_512x512_160k_ade20k.py --options checkpoint_config.max_keep_ckpts=1 dist_params.port=24727 >/dev/null &
+GPUS=4  GPUS_PER_NODE=4  CPUS_PER_TASK=2 ./tools/slurm_train.sh $PARTITION fcn_hr18s_512x512_160k_ade20k configs/hrnet/fcn_hr18s_512x512_160k_ade20k.py --options checkpoint_config.max_keep_ckpts=1 dist_params.port=24727 --work-dir work_dirs/hrnet/fcn_hr18s_512x512_160k_ade20k >/dev/null &
```
Comment (Junjun2016) [inline .dev/benchmark_train.sh:4]: Maybe we only need --work-dir `work_dirs/hrnet`.
`--work-dir work_dirs/hrnet/fcn_hr18s_512x512_160k_ade20k`
--> 
`--work-dir work_dirs/hrnet/`

File: .dev/generate_benchmark_train_script.py [hunk 8]
```diff
@@ -43,11 +44,12 @@ def create_train_bash_info(commands, config, script_name, partition, port):
         command_info = f'GPUS=4  GPUS_PER_NODE=4  ' \
                         f'CPUS_PER_TASK=2 {script_name} '
     command_info += f'{partition} '
-    command_info += f'{fname} '
+    command_info += f'{config_name} '
     command_info += f'{cfg} '
     command_info += f'--options ' \
                     f'checkpoint_config.max_keep_ckpts=1 ' \
                     f'dist_params.port={port} '
+    command_info += f'--work-dir work_dirs/{model_name}/{config_name} '
```
Comment (Junjun2016) [inline .dev/generate_benchmark_train_script.py:52]:
```suggestion
    command_info += f'--work-dir work_dirs/{model_name}'
```

Does it work?

**Review by Junjun2016 — COMMENTED** [review]
gather_model should also be modified.

**Comment by Junjun2016** [comment]
`\` --> `/`

**Review by Junjun2016 — APPROVED** [review]
LGTM

**Review by xvjiarui — APPROVED** [review]

**Pull request merged**